### PR TITLE
Add s4d

### PR DESCRIPTION
**Pull request by kyoon-mit**
## Summary

Adds the S4D state-space model architecture to `ml4gw`.

- New model in `ml4gw/nn/ssm/s4d.py`.
- Tests in `tests/nn/ssm/test_s4d.py`
- Benchmark in `benchmarks/nn/test_s4d.py`: forward pass across batch size and testing O(L) behavior for sequence length L.

## Benchmark

Forward-pass medians on a shared NVIDIA A100.

| batch | L=512 | L=2048 | L=4096 |
|------:|------:|-------:|-------:|
| 32 | 4.0 ms | 12.7 ms | 23.7 ms |
| 128 | 10.9 ms | 41.9 ms | 84.5 ms |
| 512 | 40.4 ms | 162.2 ms | 328.8 ms |

- Confirms O(L) scaling.

### CODE REVIEW

**Review by wbenoit26 — CHANGES_REQUESTED** [review]
Overall, this looks good to me. I'm not familiar with the details of what this architecture should look like, so it would be good to get a second pair of eyes from someone who is; maybe @chreissel or @benedict-armstrong could confirm.

Could you also add a couple small tests for the cases of `transpose=False`, `lr=0.0`, `lr=None`, and a test to make sure the dropout probability error gets raised cleanly?

File: ml4gw/nn/ssm/s4d.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+# Adapted with modifications from https://github.com/state-spaces/s4
+# (models/s4/s4d.py and src/models/nn/dropout.py), licensed under Apache-2.0.
+
+import math
+
+import torch
+import torch.nn as nn
+
+
+class DropoutNd(nn.Module):
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: This seems general purpose, let's move it to a `dropout.py` similar to the `norm.py` under [nn](https://github.com/ML4GW/ml4gw/tree/main/ml4gw/nn)

File: ml4gw/nn/ssm/s4d.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    """N-dimensional dropout that ties the mask across sequence positions."""
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: I don't understand the docstring. Please provide an example usage.

File: ml4gw/nn/ssm/s4d.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            if not self.transposed:
+                X = X.movedim(-1, 1)
+            mask_shape = (
+                X.shape[:2] + (1,) * (X.ndim - 2) if self.tie else X.shape
+            )
+            mask = torch.rand(*mask_shape, device=X.device) < 1.0 - self.p
+            X = X * mask * (1.0 / (1 - self.p))
+            if not self.transposed:
+                X = X.movedim(1, -1)
+        return X
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: This can be made more general purpose by avoiding `if self.training`. The assumption is whatever code calls this dropout block will only use it in the training loop.

File: ml4gw/nn/ssm/s4d.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+
+class S4DKernel(nn.Module):
+    """Generate convolution kernel from diagonal SSM parameters."""
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: Improve docstring. Describe the arguments. Also a highlight of what is being done.

File: ml4gw/nn/ssm/s4d.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        self,
+        d_model: int,
+        N: int = 64,
+        dt_min: float = 0.001,
+        dt_max: float = 0.1,
+        lr: float | None = None,
+    ):
+        super().__init__()
+
+        H = d_model
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: why not use `H` in the function signature

File: ml4gw/nn/ssm/s4d.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+        C = C * (torch.exp(dtA) - 1.0) / A
+        K = 2 * torch.einsum("hn, hnl -> hl", C, torch.exp(K)).real  # (H, L)
+        return K
+
+    def register(
+        self, name: str, tensor: torch.Tensor, lr: float | None = None
+    ) -> None:
+        """Register a tensor with a configurable LR and 0 weight decay."""
+        if lr == 0.0:
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: When is lr zero? Or in other words, can we design this so that we choose the "training" phase as the default. For the testing phase, the attribute can be changed in the test step.

File: ml4gw/nn/ssm/s4d.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        log_dt = torch.rand(H) * (
+            math.log(dt_max) - math.log(dt_min)
+        ) + math.log(dt_min)
+
+        C = torch.randn(H, N // 2, dtype=torch.cfloat)
+        self.C = nn.Parameter(torch.view_as_real(C))
+        self.register("log_dt", log_dt, lr)
+
+        log_A_real = torch.log(0.5 * torch.ones(H, N // 2))
+        A_imag = math.pi * torch.arange(N // 2).repeat(H, 1)
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py:79]:
```suggestion
        A_imag = torch.pi * torch.arange(N // 2).repeat(H, 1)
```
removes the `math` import

**Review by deepchatterjeeligo — CHANGES_REQUESTED** [review]
@kyoon-mit Overall fine. I think however this design intermingles training vs. inference modes together. That role should be kept for downstream, like the train dataloader or the test dataloader. One thing that you can do is to choose one mode as a default, like the training mode i.e. keep `register_parameter` and zero weight decay by default. Also if you can clarify what `lr=0` does.

File: benchmarks/nn/test_s4d.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+"""Benchmarks for the S4D sequence model."""
+
+import pytest
+import torch
+from constants import NUM_CHANNELS
+
+from ml4gw.nn.ssm.s4d import S4Model
+
+
+@pytest.fixture(params=[512, 2048, 8192], ids=lambda n: f"L_{n}")
```
Comment (deepchatterjeeligo) [inline benchmarks/nn/test_s4d.py]: Just so that we have a number for `8192` than OOM, should be change this to `4096`

File: ml4gw/nn/ssm/s4d.py (hunk 7 above)
Comment (deepchatterjeeligo) [reply ml4gw/nn/ssm/s4d.py:79]: actually I realize that there are other places where `math` is used. So this can be ignored.

File: ml4gw/nn/ssm/s4d.py (hunk 1 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Fixed and committed.

File: ml4gw/nn/ssm/s4d.py (hunk 2 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Replaced custom dropout function with `torch.nn.Dropout1d`

File: ml4gw/nn/ssm/s4d.py (hunk 3 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Replaced custom dropout function with `torch.nn.Dropout1d`.

File: ml4gw/nn/ssm/s4d.py (hunk 4 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Done.

File: ml4gw/nn/ssm/s4d.py (hunk 5 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: I think this was an intentional design choice by the authors. Internally, `H` (hidden dimension) gets passed around for consistency with the paper (2206.11893), but externally, `d_model` is used for clarity.

File: ml4gw/nn/ssm/s4d.py (hunk 6 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: It seems like the S4 authors left it there to experiment with which parameters can be freezed during training time (see Table 3. (b) of 2206.11893, for example). Since I don't think this is needed anymore, I removed it.

File: ml4gw/nn/ssm/s4d.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+
+class S4DKernel(nn.Module):
+    """Build the convolution kernel for one S4D layer.
+
+    Each channel is an independent diagonal linear time-invariant
+    state-space model, defined in continuous time by
+
+        x'(t) = A * x(t) + B * u(t)
+        y(t)  = C * x(t)
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: Use mathjax. See [these lines](https://github.com/ML4GW/ml4gw/blob/main/ml4gw/gw.py#L287-L292) for an example.

File: ml4gw/nn/ssm/s4d.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    with A diagonal. Linear time invariance means the output equals the
+    input convolved with a single kernel K, so the whole sequence is
+    processed in one convolution instead of stepped through position by
+    position. Discretizing with timestep dt gives the kernel this module
+    returns:
+
+        K_l = 2 * Re(
+            sum_n C_n * ((exp(dt * A_n) - 1) / A_n) * exp(dt * A_n)^l
+        ),
+        l = 0, ..., L-1   (n runs over the N/2 states)
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: Can we convert this to latex for an easier read. Also can you built the docs, and paste a screenshot of the rendered html. There are instructions in the contribution guide.

File: ml4gw/nn/ssm/s4d.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+            i.e. independent SSMs created.
+        N: State size per model dimension.
+            The states are stored as N / 2 conjugate pairs, so N must be even.
+        dt_min: Lower bound of the per-channel timestep, sampled
+            log-uniformly in [dt_min, dt_max]. The timestep sets the
+            timescale each SSM resolves.
+        dt_max: Upper bound of the per-channel timestep.
+        lr: Optional learning rate for `dt` and `A` parameters.
+            If None, these parameters are optimized with the same LR as the
+            rest of the model. If 0.0, these parameters are frozen.
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: Let's get rid of this. If zero is a proxy for freezing parameters, we can handle that downstream.

File: ml4gw/nn/ssm/s4d.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    Args:
+        d_model: Model dimension.
+            Equivalent to the number of channels in CNN nomenclature,
+            i.e. independent SSMs created.
+        N: State size per model dimension.
+            The states are stored as N / 2 conjugate pairs, so N must be even.
+        dt_min: Lower bound of the per-channel timestep, sampled
+            log-uniformly in [dt_min, dt_max]. The timestep sets the
+            timescale each SSM resolves.
+        dt_max: Upper bound of the per-channel timestep.
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py:50]: thanks for adding this

File: ml4gw/nn/ssm/s4d.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+        H = d_model  # Number of independent SSMs (one per channel).
+
+        ### SSM parameter initialization ###
+        # For each channel, draw a random timestep `dt`
+        # from a log-uniform distribution.
+        log_dt = torch.rand(H) * (
+            math.log(dt_max) - math.log(dt_min)
+        ) + math.log(dt_min)
+        self.register("log_dt", log_dt, lr)
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py]: Sorry if I was unclear last time. What I meant is that let's get rid of `.register` method altogether. Instead this line will be

```suggestion
        self.register_parameter("log_dt", log_dt)
```
Setting the weight decay to zero for `log_dt` will be done as a part of the `configure_optimizer` method of a lightningmodule. See [this function](https://github.com/state-spaces/s4/blob/main/example.py#L247-L291) in s4d repo, and in particular [these lines](https://github.com/state-spaces/s4/blob/main/example.py#L268-L276) where the `_optim` for certain parameters are handled.

File: ml4gw/nn/ssm/s4d.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+            y = y.transpose(-1, -2)
+        return y, None
+
+
+class S4Model(nn.Module):
+    """Full S4D sequence model for regression / classification.
+
+    Input:  (B, d_input, L)  — channels-first (aframe convention).
+    Output: (B, d_output)
+    """
```
Comment (deepchatterjeeligo) [inline ml4gw/nn/ssm/s4d.py:203]: If this is the final usable model, let's add an example in the docstring for the usage.

**Review by deepchatterjeeligo — CHANGES_REQUESTED** [review]
Hi @kyoon-mit I think this is good progress. But maybe I was not clear about the `.register` last time. Could you take another pass. There are some minor comments about the docs as well.

File: ml4gw/nn/ssm/s4d.py (hunk 11 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Fixed in bc5c88e836ed97e725205808e2941ebc67e832d3.

File: ml4gw/nn/ssm/s4d.py (hunk 13 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Fixed in bc5c88e836ed97e725205808e2941ebc67e832d3.

File: benchmarks/nn/test_s4d.py (hunk 8 above)
Comment (kyoon-mit) [reply benchmarks/nn/test_s4d.py]: Fixed in 61ed6560874ac2ed83b71a207c6c0c20d8248b23. Also, the table is updated in the initial PR comment.

File: ml4gw/nn/ssm/s4d.py (hunk 10 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: image

Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py]: Please see c0720658287af999514632a4b2bd07750036ff0a for updates to docstrings.

File: ml4gw/nn/ssm/s4d.py (hunk 14 above)
Comment (kyoon-mit) [reply ml4gw/nn/ssm/s4d.py:203]: Example added in c0720658287af999514632a4b2bd07750036ff0a.

**Comment by kyoon-mit** [comment]
@wbenoit26 The optional `lr` parameter is now gone. The suggested dropout guard and the `transpose=False` test are added in bd8b67728a0d924dcd46311326ba2f9845e3c2d6.

Final version in c1403f7b359f135973cc8f4967b562bcde5a757e (removed meaningless None returns from the forward method).

**Review by deepchatterjeeligo — APPROVED** [review]
LGTM

**Review by wbenoit26 — APPROVED** [review]
